### ISSUE TITLE
Complete description of the available dynamic modules

### ISSUE THREAD

**Issue by mariusrak** [meteor/meteor#10767]
> The initial JS bundle must include a complete description of the available dynamic modules and their dependencies to avoid multiple nested round-trips when requesting dynamic modules. This abstract description of the dynamic graph (module paths, hashes, and dependencies) is embedded in the initial JS bundle alongside the code for non-dynamic modules.

I'm sorry, but I don't understand how this is preventing multiple nested round-trips, or how omitting those dirs and files would cause the multiple nested round trips. I would also like to get rid of majority of the content of the `app.js` file. Mostly because I have an application that is meant to contain many large modules and single user would need to load only minority of them.  Now it is required to provide every user with structure of every module in the application. 

Is there no way to include only statically bundled files, that are included using `import '...'` and omit those that are included using dynamic import `import()`? When dynamically loaded file is fetched it would "install" all its dependencies into the tree. As I played around with the fetching I got stopped at the fact that if a file is not mentioned in the tree in the `app.js`, the fetching function would just throw an error. What would happen if it didn't throw an error but go to server with the request instead? I wanted to try but it was too complicated for me to edit the meteor files.

This would also open new possibilities to control fetching and also allow for more dynamic fetching with something like `import('folder/'+someComputedSubFolder+'/file')`. And maybe even allow to absolutely hide some modules that a user should not have access to.

Thanks for an explanation.

_Originally posted by @mariusrak in https://github.com/meteor/meteor/issues/10532#issuecomment-551334836_

Sorry for new issue, but I guess, nobody would get to the closed one and I find this issue to be quite a problem.

**Comment by benjamn** [comment]
With a complete client-side description of the dependency graph, including both static and dynamic modules, we avoid the problem of having to (dynamically) fetch and receive a module before finding out what its dependencies are, which delays fetching those dependencies, which delays finding out what their dependencies are, which delays fetching their transitive dependencies, and so on. This is called the _waterfall problem_, and Meteor solves it by providing just enough information for the client to understand the entire dependency graph without help from the server.

More importantly, the description contains the hashes of all available modules, so the client can check an IndexedDB cache for each module/hash before requesting it from the network, which allows the client to avoid requesting the same version of any particular module more than once.

This is a fundamentally more powerful caching model than you will find in any other module system. Progressive web apps come close, by requesting and caching modules individually (rather than bundling them together), but they pay a large up-front cost due to the waterfall problem described above.

Unlike other module systems, which compile away the module identifier strings that are used to import modules, Meteor *does* support arbitrary `import(<expression>)` calls, as long as the requested module was known about at build time, but you are responsible for including the equivalent `import("some/string/literal")` call somewhere else in your code. Otherwise, the module will be missing from the description of the dependency graph, and cannot be requested using Meteor's dynamic import system. This is a valuable limitation, because it limits client access to server files that you intentionally imported in your source code, rather than granting access to any available file within your application. You could imagine configuring these restrictions in some sort of configuration file, but you can accomplish the same thing right now, just by using `import` and dynamic `import()` within a Meteor application.

In other words, having a single description of all available modules is exactly how we "hide some modules that a user should not have access to."

**Comment by mariusrak** [comment]
Wouldn't it be possible to split whole tree by some boundaries? I have in mind the [image from react laodable article](https://miro.medium.com/proxy/1*ODrHocx1NxGmuwXq6f57xw.png). The shaded parts would represent the boundaries. This way on first load only some initial module (with its tree) would be loaded. Next module would of course be delayed and make a waterfall, but it would provide its part of tree, which could merge with already loaded tree. The subtree of second loaded module should contain all required paths. I see, that here would have to be included also some dependencies that were already loaded with initial load, but would it be that much overhead? After the subtree loading and merging into already loaded tree, everything can continue as it is done nowadays.

So basically, only change would be in step where currently the `require` throws an error. The change would be, that if a file is not in a tree, the client would require subtree for the module. Server would reply (e.g.) with tree for the requested file and dependencies imported by the requested file. Paths can start from root for simpler merge. After fetching the subtree, client can continue fetching dynamically imported files as nowadays. I see the problem with current implementation only in this part, that instead of some fallback to server, the `require` function throws an error. But also tree generating would have to be rethought, I guess.

>Meteor does support arbitrary import(<expression>) calls, as long as the requested module was known about at build time. However, you are responsible for including the equivalent import("some/string/literal") call somewhere else in your code.  

I like how it is currently done with arbitrary `import(<expression>)` calls. I do it this way myself, I just thought it's a hack :D 

The way I propose, everything would be secured same as it is now. I.e. only intentionally imported files would be included in build and only those would be provided by server to fetch. However, one could implement `WebApp` middleware which could respond even to files that were not found/included and e.g. build them on-the-fly. Or the middleware could also respond to files that are in the build and restrict access. This would provide new level of flexibility.

**Comment by mariusrak** [comment]
Also to back-up my view: imagine a system with some sort of simple front end, e.g. to display statistics to wide public. The system would also contain some restricted portal, which would be available for registered users and would provide various useful tools which would be complex. The system would also contain complex back-end for few administrators. By complex I mean also complex ui features, which can grow quite large. All the parts would share some functionalities, server and database, but also a lot of parts would be different for each of the groups.

With current approach, all 3 parts would have to be served to all 3 groups of users.. But the wide public would not need all the dependencies which comes with editing possibilites of top-level administrator. While it is possible to deploy 3 systems at one server, make some reusable libraries that would be included in all of them and connect them all to single database, I think that it is quite unnecessary burden. Same as I think is not necessary to download whole tree for all of the visitors.

**Comment by benjamn** [comment]
The `require` function cannot have a server-based fallback behavior, because it's synchronous. Only dynamic `import()` has the flexibility of fetching modules from the server if they're not available locally, so dynamic `import()` would be the only place we could split the dependency graph.

The drawback of splitting the dependency graph is that the client has to wait for the additional dependency graph information to arrive before checking its IndexedDB cache of modules, and only then can it request the missing modules from the server. Instead, Meteor's current implementation bundles the entire dependency graph (which is much smaller than the actual modules) into the initial JS bundle, a relatively small incremental cost that allows all future dynamic module fetching to happen in a single HTTP request.

I still think these are the right tradeoffs, all things considered. Specifically, I do not think performance would be improved by waiting to request subgraphs of the dependency tree until a dynamic `import()` happens. I could imagine a system where those subgraphs arrive shortly after the initial bundle, so that they are likely to be available before they are needed, or a system where the client reports a hash of the last graph it received, so the server can send down some sort of diff instead of an entirely new graph. The barrier to both of those ideas is that they are more complex, so the performance benefits would need to be very tangible, not just theoretical.

**Comment by mariusrak** [comment]
> The `require` function cannot have a server-based fallback behavior, because it's synchronous. Only dynamic import() has the flexibility ...

I'm currently not sure how it's implemented, but AFAIK when I require a module at client the tree is checked and if the module is in the tree but has no code (i.e. is not loaded), then client fetches the code. But if the module is not in the tree it fails. So this place where, the module is checked and fetched could implement one more fall back. It would be asynchronous same as fetching the `import()`-ed module.

> The drawback of splitting the dependency graph is that the client has to wait for the additional dependency graph information to arrive before checking its IndexedDB cache of modules

Maybe the tree can be also cached? There is IMO no problem if I have some parts of tree loaded some time ago and I visit the system and the parts are still there. I only don't want to burden all client with all code. But if a part of a tree was relevant for me in past, it would possibly be relevant also in present.

> I do not think performance would be improved by waiting to request subgraphs of the dependency tree until a dynamic import() happens.

Of course in time of loading the dynamic module, the load would take longer, the question is how much longer. But on the other side we would get shorter initial load which can provide an animation to ease the loading. So while performance might be worse by numbers, it would seem to be better for users.

So we can narrow it down to tradeoff between (A) one more HTTP request per dynamic import - increasing loading time for each dynamically imported module, and (B) complete a therefore larger dependency graph - increasing initial load in case of large system. Do I understand it correctly? 

Of course there is no "one size fits it all" but I think this is worth considering and maybe wider discussion or some test.

**Comment by afrokick** [comment]
Meteor uses WebSocket, so the loading of dynamic import is very fast without any handshakes/etc.

**Comment by mariusrak** [comment]
AFAIK dynamic modules are loaded by HTTP. Each fetch is done in a new HTTP request. But maybe the trees could be loaded by WebSocet which would lower the delay.

**Comment by benjamn** [comment]
The original implementation in Meteor 1.5 did use a WebSocket, but we switched to HTTP during the 1.6.x release process, since you can kick off HTTP requests earlier, without loading SockJS first.

**Comment by benjamn** [comment]
> So we can narrow it down to tradeoff between (A) one more HTTP request per dynamic import - increasing loading time for each dynamically imported module, and (B) complete a therefore larger dependency graph - increasing initial load in case of large system. Do I understand it correctly?

That's a good way of framing it, yes. I'm strongly in favor of B, since I don't think initial load times would be improved by removing the dynamic module graph from the initial JS bundle, whereas dynamic module fetches would definitely take longer without complete information about dynamic modules. Reducing the number of HTTP requests (of any size) tends to have a bigger impact than reducing the size of a single resource by ~5%.

We can keep brainstorming ways to solve both problems at once, but I hope I've convinced you that the current design makes reasonable tradeoffs in most cases.

**Comment by mariusrak** [comment]
> I don't think initial load times would be noticeably improved by removing the dynamic module graph from the initial JS bundle

Can we search for some numbers? I have single (currently just) experimental application which might not be best sample. I believe you have access to some projects which heavily uses the dynamic modules. It might help to look at the bundle, remove paths without code and then compare its size to original size. I'd be happy to help if we'd had some list of large meteor based applications.

> dynamic module fetches would definitely take longer without complete information about dynamic modules

Just to remind, what is the required information about the modules? Is it only the paths, i.e. the subtree (with hashes I guess, even though I cannot see them in my bundle) or is there anything more needed before fetch?

> Reducing the number of HTTP requests (of any size) tends to have a bigger impact 

Would it be possible to load the tree by WebSocket? This would keep the HTTP request count same.

Or, currently, to load modules a subtree is sent to server with all paths to load, right? But is this even correct way? When client wants to load something it is loaded by single function call. Coded as `import(...)` transpiled to `dynamicImport(...)`, I think. But the subtree is an information that the server already has, not? So what if only the string from argument was sent to server and server resolved which dependencies to send down? Would that be possible? This would even reduce the fetch request size. Or what is a reason for sending the subtree from client instead of only entry point? How is the subtree to fetch calculated at the client?

> We can keep brainstorming ways to solve both problems at once

If there would be no way to go by the A path, then maybe some configuration or fork of builder package that would do it the other way?

**Comment by mariusrak** [comment]
Can we please discuss it further? It is almost critical issue for me and I believe it can be valuable for others also.

**Comment by red-meadow** [comment]
> So what if only the string from argument was sent to server and server resolved which dependencies to send down? Would that be possible? This would even reduce the fetch request size. Or what is a reason for sending the subtree from client instead of only entry point?

I guess the reason is that only client nows which modules it has cached and which ones it has to fetch.

**Comment by red-meadow** [comment]
One quick thought after reading the discussion. The dependency graph contains full names of folders and components right now, which are the main source of additional KBs. Maybe we could minify it? What if instead of sending
```javascript
imports: {
  db: {
    'collection.js': []
  },
  api: {
    'service.js': ['/imports/db/collection.js']
  },
  client: {
    'component.js': ['/imports/db/collection.js', '/imports/client/service.js']
  }
}
```
we provide some server-side path mapping, like
```javascript
[
  a: '/imports/db/collection.js',
  b: '/imports/client/service.js',
  c: '/imports/client/component.js'
]
```
and send a minified version to the client:
```javascript
{
  a: [],
  b: ['a'],
  c: ['a', 'b']
}
```
It will definitely break this code: `import(<expression>)`. Also, having relative paths seems to provide an additional layer of complexity as they should be mapped to the same key. Am I missing something else?

**Comment by red-meadow** [comment]
@benjamn, you mentioned here https://github.com/meteor/meteor/pull/10811#issuecomment-564726713, that you still have big plans for the future of Meteor's module system. Could you describe very briefly the direction of possible changes? Do you think that in the foreseeable future we can switch to using native ES modules (and dynamic imports) with no bundling in modern browsers?

**Comment by mariusrak** [comment]
I'm not looking for mitigation, but for scalable solution. The proposal with paths minification only mitigates the problem by some factor. But in case of extremely large projects the problem will be same. And i guess, it is not only about bytes but also about memory usage and objects complexity. And also about looking for the right solution.

And the right solution (imo) doesn't download whole project file map to all clients, but only files relevant for the client.

**Comment by red-meadow** [comment]
Still relevant

**Comment by jmdsg** [comment]
I'm having the same issue here.
I think that the current solution is overkill. I now that using gzip, and having this data repeated (a lot), makes the final js to have a lower entropy, and the compression rate to be higher. But still the bundle size becomes way to big for larger projects.

The current solution makes that every since file in the app, that apparently can be lazily imported (including the complete subtree), to have a version hash of 40bytes each. (Just that represents more than 100KB in my app).

And then the entire app map, with all the imports in the app (raw, as you import them). The bundle size can even increase when you split a file in two, or when you use large file names, that are used as import many times.

I am using ramda in my project, and decided to use the babel plugin to use direct imports, the bundle size of ramda went from 100KB to 50KB, but then, since every single import is bundled and repeated, i ended up with hundreds and hundreds of imports like ramda/src/identity, ramda/src/map..., and my bundle size increased more than the 50KB than i saved.

I don't think it would be easy to use some kind of scope. But at least, using a lookup table as sugested by @red-meadow, can be a good start. And probably this would make necessary to use an absolute path for local files, instead of the relative one. I am also concerned by the dynamic versions.

**Comment by jmdsg** [comment]
Just to give an example, i mean something like:

```
// dynamic-versions.js
var deps = ["ramda", "/imports/absolute/to/my/path", ...];
var versions = ["Here+20+Base64+Chars", "...", ...];

// app map
{
  folder: {
    "file.js": [0, 3, 6], // deps
  }
}
```

It easier to handle, just access by index. And if not all the deps are versioned, then another array indicating which ones are versioned, can be used `var versioned = [1, 3];`

**Comment by megawebmaster** [comment]
I just stumbled upon this issue and I might shed an interesting point of view for that. My app currently has 120kB of this dynamic imports descriptors (using Meteor 2.3.5) for the initial page when I dynamically load only 2 components with a multiple dynamic dependencies (there are different routes there).

I believe that the bundling system and dynamic imports should support what developer wants to do - split at the `import()` statements. To make it work we should send a set of dynamic imports, but do *not* send anything we find on the next level of `import()` causing code to split in a place where developer wants it.

Example:

```jsx
// app.jsx
const Page1 = lazy(() => import('./page1.jsx'));
const Page2 = lazy(() => import('./page2.jsx'));

export const App = (abTest) => (
  <Suspense fallback={<div>Loading...</div>}>
    {abTest ? <Page1 /> : <Page2 />}
  </Suspense>
);

// page1.jsx
const Subpage1 = lazy(() => import('./page1/subpage1.jsx'));
const Subpage2 = lazy(() => import('./page1/subpage2.jsx'));

const Page1 = () => (
  <Router>
    <Switch>
      <Suspense fallback={<div>Loading...</div>}>
        <Route path="/" exact component={Subpage1} />
        <Route path="/sub" component={Subpage2} />
      </Suspense>
    </Switch>
  </Router>
);

export default Page1;

// page2.jsx
const Subpage1 = lazy(() => import('./page2/subpage1.jsx'));
const Subpage2 = lazy(() => import('./page2/subpage2.jsx'));

const Page2 = () => (
  <Router>
    <Switch>
      <Suspense fallback={<div>Loading...</div>}>
        <Route path="/" exact component={Subpage1} />
        <Route path="/sub" component={Subpage2} />
      </Suspense>
    </Switch>
  </Router>
);

export default Page2;
```

In this case NOT sending the whole tree (about any of the `Subpage`s) is a good thing because as a user I don't have to download unnecessary data. First page load should deliver data to see whether we need to download `page1.jsx` or `page2.jsx`. If we have correct hashes in our cache then we don't even need to download anything else if we could make hashes include dependent files caches (for example). This increases how complex the build system is, but it benefits the end user greatly.

Seeing what Facebook does with Relay and other technologies it's the way industry goes - optimizing what the user receives (at the expense of server computing time and a bit of networking if we don't have a SSR-enabled site) to improve performance and perceived speed. I also think that HTTP/2 (and soon: QUIC) improves how fast we can make requests and it's not the right solution to send all the details down to the client.

What do you think about all that?
